### DISCUSSION TITLE
writeln() does not write a new line and every write is shown twice and i need to click on the div to update the text

### DISCUSSION THREAD

**Discussion by JensvandeWiel** [sourcelair/xterm.js#4615] [Q&A] [answered]
*Edit: in production it works fine
## Details
- Browser and browser version: microsoft webview2 / wails (electron alternative)
- OS version: windows 10 22h2 pro
- xterm.js version: 5.2.1

### Steps to reproduce

1. create following react component with mui/joy
```js
import {Component} from "react";
import React from "react";
import {Alert, Button, Divider, LinearProgress, styled, Theme, useTheme} from "@mui/joy";
import WarningAmberRoundedIcon from '@mui/icons-material/WarningAmberRounded';
import '../../../wailsjs/runtime/runtime';
import {EventsOff, EventsOn, LogPrint} from "../../../wailsjs/runtime";
import {Install} from '../../../wailsjs/go/installer/Installer';
import {Terminal} from "xterm";
import 'xterm/css/xterm.css';


//ts
interface State {
    buttonVisible: boolean;
}



export class Installer extends Component<{}, State> {
    //PROPS AND STATE
    private term: Terminal = new Terminal();


    constructor(props: {}) {
        super(props);
        this.state = {
            buttonVisible: true,
        };
    }

    // region EVENTS

    componentDidMount() {
        EventsOn("printToConsole", (line: string) => {
            this.WriteLineToLog(line);
        });

        const element = document.getElementById('console') as HTMLElement;

        this.term.open(element);
        this.term.writeln('Hello from \x1B[1;3;31mxterm.js\x1B[0m $ \r\n');
        this.term.writeln('Hello from \x1B[1;3;31mxterm.js\x1B[0m $ \r\n');



    }

    componentWillUnmount() {
        // Cleanup the event listener here when the component is unmounted
        EventsOff("printToConsole");

        if (this.term) {
            this.term.dispose();
        }
    }

    WriteLineToLog(line: string){

    }

    handleButtonClick = () => {
        if (this.term) {
            this.term.writeln('New line written on button click!\r\n\n\n'); // Add a new line when the button is clicked
        }
    };

    //endregion

    render() {
        const Level1Div = styled('div')(({ theme }) => ({
            // The result is 'var(--joy-palette-primary-500)'
            backgroundColor: theme.vars.palette.background.level1,
        }));
        return (
            <>
                <Level1Div className="">
                    <LinearProgress value={40} className="mt-1" />
                    <p className="text-center text-xl mt-2">Installing Rust server </p>
                    <Divider orientation="horizontal"/>
                    <p className="font-bold ">Log:</p>
                    <Button onClick={this.handleButtonClick} ></Button>
                    <div className="" id="console"></div>
                </Level1Div>

            </>
        );
    }
}
```
2. Run and check

I've been messing with it for 3 hourse now, can't find the problem
Maybe its something stupid or a option but i can seem to find out.
Please help 🙏

**Comment by JensvandeWiel** [comment]
![image](https://github.com/xtermjs/xterm.js/assets/85284773/a9bf6892-6feb-4c2f-aed8-cf5dc2a71312)

**Comment by JensvandeWiel** [comment]
Ok, i found out that the double printing is because its being called in componentDidMount without that and just clicking the button a couple of times i get:
![image](https://github.com/xtermjs/xterm.js/assets/85284773/47ebc9de-ae54-4d07-bf93-0053db4c6262)

**Comment by Tyriar** [comment]
Try setting options.logLevel to `'debug'` to see what is being written exactly to xterm.js

**Reply by JensvandeWiel** [reply]
ill give that a try 👍

**Reply by JensvandeWiel** [reply]
when i click the button twice:
![image](https://github.com/xtermjs/xterm.js/assets/85284773/a1f1b875-eaff-48d3-b00e-a2723ae41cad)
does not look to good

**Reply by Tyriar** [reply]
@jerch have you seen this Unknown EXECUTE code log for \n and \r before? 😕

**Reply by JensvandeWiel** [reply]
May help, but before this i had a project that catured console output in a pseudo console and it gave the same codes then

**Reply by jerch** [reply]
:scream: Nope, this should never happen for a standard terminal instance (means w'o messing yourself with parser handlers), CR & LF handlers should be attached right from the start (by the inputhandler ctor).

**Reply by JensvandeWiel** [reply]
Im trying to use it as barebones as possible :(

**Reply by Tyriar** [reply]
@JensvandeWiel if you clone the repo and [run the demo](https://github.com/xtermjs/xterm.js/wiki/Contributing#running-the-demo) do you see the same thing? My guess is react and/or your bundler is messing with something.

**Reply by jerch** [reply]
Can you create a minimal repro project and link to it here? If the default CR & LF handlers are gone, then something is seriously broken.

**Reply by JensvandeWiel** [reply]
@jerch The same type of program or a normal react program?

**Reply by JensvandeWiel** [reply]
in case here is a link to the repo itself: https://github.com/JensvandeWiel/RustManager
file: https://github.com/JensvandeWiel/RustManager/blob/dev/frontend/src/Tabs/Serverconfig/Installer.tsx

**Reply by jerch** [reply]
@JensvandeWiel Idc, something that can be bootstrapped within a minute and does not pull half of all npm packages out there.

**Reply by JensvandeWiel** [reply]
@jerch my project right now uses not much npm packages if you clone it and run wails dev in the base directory is should work
otherwise here is a test repo but it does not use wails so i do not know if it would do the same:
https://github.com/JensvandeWiel/tetst

**Reply by JensvandeWiel** [reply]
@Tyriar if i try to run the example i get:
PS C:\Users\Jens\Downloads\xterm.js> npm run setup

> xterm@5.2.1 presetup
> node ./bin/install-addons.js

pulling addon dependencies...
Skipped xterm-addon-attach
Skipped xterm-addon-canvas
Skipped xterm-addon-fit
Preparing xterm-addon-ligatures
error An unexpected error occurred: "Invalid value type 41:0 in C:\\Users\\Jens\\Downloads\\xterm.js\\addons\\xterm-addon-ligatures\\yarn.lock".
node:child_process:960
    throw err;
    ^

Error: Command failed: yarn
error An unexpected error occurred: "Invalid value type 41:0 in C:\\Users\\Jens\\Downloads\\xterm.js\\addons\\xterm-addon-ligatures\\yarn.lock".

    at checkExecSyncError (node:child_process:885:11)
    at Object.execSync (node:child_process:957:15)
    at C:\Users\Jens\Downloads\xterm.js\bin\install-addons.js:47:14
    at FSReqCallback.oncomplete (node:fs:198:23) {
  status: 1,
  signal: null,
  output: [
    null,
    Buffer(289) [Uint8Array] [
      121,  97, 114, 110,  32, 105, 110, 115, 116,  97, 108, 108,
       32, 118,  49,  46,  50,  50,  46,  49,  57,  10, 105, 110,
      102, 111,  32,  73, 102,  32, 121, 111, 117,  32, 116, 104,
      105, 110, 107,  32, 116, 104, 105, 115,  32, 105, 115,  32,
       97,  32,  98, 117, 103,  44,  32, 112, 108, 101,  97, 115,
      101,  32, 111, 112, 101, 110,  32,  97,  32,  98, 117, 103,
       32, 114, 101, 112, 111, 114, 116,  32, 119, 105, 116, 104,
       32, 116, 104, 101,  32, 105, 110, 102, 111, 114, 109,  97,
      116, 105, 111, 110,
      ... 189 more items
    ],
    Buffer(145) [Uint8Array] [
      101, 114, 114, 111, 114,  32,  65, 110,  32, 117, 110, 101,
      120, 112, 101,  99, 116, 101, 100,  32, 101, 114, 114, 111,
      114,  32, 111,  99,  99, 117, 114, 114, 101, 100,  58,  32,
       34,  73, 110, 118,  97, 108, 105, 100,  32, 118,  97, 108,
      117, 101,  32, 116, 121, 112, 101,  32,  52,  49,  58,  48,
       32, 105, 110,  32,  67,  58,  92,  92,  85, 115, 101, 114,
      115,  92,  92,  74, 101, 110, 115,  92,  92,  68, 111, 119,
      110, 108, 111,  97, 100, 115,  92,  92, 120, 116, 101, 114,
      109,  46, 106, 115,
      ... 45 more items
    ]
  ],
  pid: 1352,
  stdout: Buffer(289) [Uint8Array] [
    121,  97, 114, 110,  32, 105, 110, 115, 116,  97, 108, 108,
     32, 118,  49,  46,  50,  50,  46,  49,  57,  10, 105, 110,
    102, 111,  32,  73, 102,  32, 121, 111, 117,  32, 116, 104,
    105, 110, 107,  32, 116, 104, 105, 115,  32, 105, 115,  32,
     97,  32,  98, 117, 103,  44,  32, 112, 108, 101,  97, 115,
    101,  32, 111, 112, 101, 110,  32,  97,  32,  98, 117, 103,
     32, 114, 101, 112, 111, 114, 116,  32, 119, 105, 116, 104,
     32, 116, 104, 101,  32, 105, 110, 102, 111, 114, 109,  97,
    116, 105, 111, 110,
    ... 189 more items
  ],
  stderr: Buffer(145) [Uint8Array] [
    101, 114, 114, 111, 114,  32,  65, 110,  32, 117, 110, 101,
    120, 112, 101,  99, 116, 101, 100,  32, 101, 114, 114, 111,
    114,  32, 111,  99,  99, 117, 114, 114, 101, 100,  58,  32,
     34,  73, 110, 118,  97, 108, 105, 100,  32, 118,  97, 108,
    117, 101,  32, 116, 121, 112, 101,  32,  52,  49,  58,  48,
     32, 105, 110,  32,  67,  58,  92,  92,  85, 115, 101, 114,
    115,  92,  92,  74, 101, 110, 115,  92,  92,  68, 111, 119,
    110, 108, 111,  97, 100, 115,  92,  92, 120, 116, 101, 114,
    109,  46, 106, 115,
    ... 45 more items
  ]
}

Node.js v18.16.0

This is the output of npm start

**Reply by JensvandeWiel** [reply]
the wails dev command just runs npm run dev wich just runs vite
vite.config.js:
```js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()]
})
```

**Reply by jerch** [reply]
This gets a little spammy now.

> ...so i do not know if it would do the same

Well I am not going to fix your toolchain/project setup for you. All I offer is to look into the reason, why the default handlers got lost on an otherwise standard terminal instance, and for that I'd need a minimal repro and not something else, that may or may not show the faulty behavior. Thats just a waste of time on my end.

**Reply by Tyriar** [reply]
> PS C:\Users\Jens\Downloads\xterm.js> npm run setup

@JensvandeWiel you want to run `yarn` to install deps

**Reply by JensvandeWiel** [reply]
@jerch As i said before its in the difference of wanneer ls build and wails dev so is probably on the framework side ive made a ticket there and now i need to wait for an response

**Reply by JensvandeWiel** [reply]
> > PS C:\Users\Jens\Downloads\xterm.js> npm run setup
> 
> @JensvandeWiel you want to run `yarn` to install deps

Yes ive tried that

**Comment by JensvandeWiel** [comment]
New news! if i just build my project for production it works fine?

![image](https://github.com/xtermjs/xterm.js/assets/85284773/435fae77-e3ce-47ec-9d96-1cba10f00aa3)
[RustManager.zip](https://github.com/xtermjs/xterm.js/files/12185284/RustManager.zip)

**Reply by Tyriar** [reply]
Good to hear, looks like it's an issue with your bundler then 🙂

**Reply by JensvandeWiel** [reply]
@Tyriar one small question do you know if xterm.js writes anything to local storage. Because i had a problem with local storage with mui and wails maybe its the same problem?

**Reply by Tyriar** [reply]
No we don't touch the locale storage api

**Comment by JensvandeWiel** [comment, marked as answer]
Update it kind of was not xterm but react, ill explain:
Dont know if this is just a stupid error on my side but:
this is the old code:
```tsx
export class Installer extends Component<{}, State> {
    //PROPS AND STATE
    private term: Terminal = new Terminal();


    constructor(props: {}) {
        super(props);
        this.state = {
            buttonVisible: true,
        };
    }

    // region EVENTS

    componentDidMount() {
        EventsOn("printToConsole", (line: string) => {
            this.WriteLineToLog(line);
        });

        const element = document.getElementById('console') as HTMLElement;

        this.term.open(element);
        this.term.writeln('Hello from \x1B[1;3;31mxterm.js\x1B[0m $ \r\n');
        this.term.writeln('Hello from \x1B[1;3;31mxterm.js\x1B[0m $ \r\n');



    }

    componentWillUnmount() {
        // Cleanup the event listener here when the component is unmounted
        EventsOff("printToConsole");

        if (this.term) {
            this.term.dispose();
        }
    }

    WriteLineToLog(line: string){

    }

    handleButtonClick = () => {
        if (this.term) {
            this.term.writeln('New line written on button click!\r\n\n\n'); // Add a new line when the button is clicked
        }
    };

    //endregion

    render() {
        const Level1Div = styled('div')(({ theme }) => ({
            // The result is 'var(--joy-palette-primary-500)'
            backgroundColor: theme.vars.palette.background.level1,
        }));
        return (
            <>
                <Level1Div className="">
                    <LinearProgress value={40} className="mt-1" />
                    <p className="text-center text-xl mt-2">Installing Rust server </p>
                    <Divider orientation="horizontal"/>
                    <p className="font-bold ">Log:</p>
                    <Button onClick={this.handleButtonClick} ></Button>
                    <div className="" id="console"></div>
                </Level1Div>

            </>
        );
    }
}
```
The terminal instance gets created outside of the constructor and componentDidMount like:
```tsx
export class Installer extends Component<{}, State> {
    //PROPS AND STATE
    private term: Terminal = new Terminal();
    ........
```
This does not work, the following does work:
```tsx
export class Installer extends Component {
    //don't make a new terminal instance here because it will break the terminal
    terminalRef: Terminal | null = null;

    componentDidMount() {
        // Handle events
        EventsOn('printToConsole', (line: string) => {
            //this.WriteLineToLog(line);
        });

        // Create a new terminal instance
        const term = new Terminal({ logLevel: 'debug', windowsMode: true });
        this.terminalRef = term;

        const termEle = document.getElementById('terminal');

        // Attach the terminal to the DOM element
        if (termEle) {
            term.open(termEle);
        }

        // Write initial text to the terminal
        term.writeln('Hello from xterm.js $');
        term.writeln('Hello from xterm.js $');
    }

    componentWillUnmount() {
        // Clean up the terminal instance and event listeners when the component unmounts
        const term = this.terminalRef;
        if (term) {
            term.dispose();
        }
        // Cleanup events
        EventsOff('printToConsole');
    }

    handleButtonClick = () => {
        const term = this.terminalRef;
        if (term) {
            term.writeln('New line written on button click!'); // Add a new line when the button is clicked
        }
    };

    render() {
        const Level1Div = styled('div')(({ theme }) => ({
            // The result is 'var(--joy-palette-primary-500)'
            backgroundColor: theme.vars.palette.background.level1,
        }));

        return (
            <>
                <Level1Div className="">
                    <LinearProgress value={40} className="mt-1" />
                    <p className="text-center text-xl mt-2">Installing Rust server</p>
                    <Divider orientation="horizontal" />
                    <p className="font-bold">Log:</p>
                    <Button onClick={this.handleButtonClick}></Button>
                    <div className="" id="terminal"></div>
                </Level1Div>
            </>
        );
    }
}

export default Installer;
```
So the terminal instance must be created after the component mounted.
Function component that works too:
```tsx
const Installer = () => {
    const terminalRef = useRef<Terminal>();
    const [buttonVisible, setButtonVisible] = React.useState(true);

    useEffect(() => {
        //handle events
        EventsOn("printToConsole", (line: string) => {
            //this.WriteLineToLog(line);
        });



        // Create a new terminal instance
        const term = new Terminal({logLevel : 'debug', windowsMode: true});
        terminalRef.current = term;

        const termele = document.getElementById("terminal");

        // Attach the terminal to the DOM element
        if (termele) {
            term.open(termele);
        }

        // Write initial text to the terminal
        term.writeln('Hello from xterm.js $');
        term.writeln('Hello from xterm.js $');


        //componentwillunmount
        return () => {
            // Clean up the terminal instance when the component unmounts
            term.dispose();
            //cleanup events
            EventsOff("printToConsole");
        };
    }, []);

    const Level1Div = styled('div')(({ theme }) => ({
        // The result is 'var(--joy-palette-primary-500)'
        backgroundColor: theme.vars.palette.background.level1,
    }));


    const handleButtonClick = () => {
        const term = terminalRef.current;
        if (term) {
            term.writeln("New line written on button click!"); // Add a new line when the button is clicked
        }
    };

    return (
        <>
            <Level1Div className="">
                <LinearProgress value={40} className="mt-1" />
                <p className="text-center text-xl mt-2">Installing Rust server </p>
                <Divider orientation="horizontal"/>
                <p className="font-bold ">Log:</p>
                <Button onClick={handleButtonClick} ></Button>
                <div className="" id="terminal"></div>
            </Level1Div>

        </>
    );
};

export default Installer;
```
It took me a long time to find the real answer but i'm happy now :)

**Reply by jerch** [reply]
Thx for providing this snippet for react. We have many questions regarding react & xtermjs, but none of us uses react. So this is def. helpful for many peeps.

**Reply by JensvandeWiel** [reply]
No problem, i'm still quite new to react so its probably not the best code but i think it will help people overall with using xtermjs with react and if they ever get this problem.
